### PR TITLE
feat: improve types to fix JSON schema when schema is passed like object with headers

### PR DESCRIPTION
**Pull request by B2o5T**
improve types and non-breaking refactoring

fixes https://github.com/kamilkisiela/graphql-config/issues/765

### CODE REVIEW

File: config-schema.json [hunk 1]
```diff
@@ -121,23 +125,25 @@
...
+          "title": "projects"
         }
       },
       "required": ["projects"],
       "title": "IGraphQLProjects",
       "type": "object"
+    },
+    "Record<string,IGraphQLProject|IGraphQLProjectLegacy>": {
+      "title": "Record<string,IGraphQLProject|IGraphQLProjectLegacy>",
+      "type": "object"
```
Comment (ardatan) [inline config-schema.json:138]: I think this should have a dynamic property like;
```
            "patternProperties": {
                ".*": {
                    "anyOf": [
                        { "$ref": "IGraphQLProject" },
                        { "$ref": "IGraphQLProject" }
                     ]
              }
 ```
 See https://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties

Comment (ardatan) [reply config-schema.json:138]: Maybe somethign like this;
https://github.com/kamilkisiela/graphql-config/pull/1011/files#diff-df4af8dbe3d6fbc388b8276ca4494407f03a6723884ee3d1f7e03f754aa241beL124

Comment (B2o5T) [reply config-schema.json:138]: this file was generated by `typescript-json-schema` package

Comment (B2o5T) [reply config-schema.json:138]: @ardatan I think it's not supported by `typescript-json-schema`
Tried to generate the following types
image
Received the following error
```
/Users/dimitri/Desktop/GUILD/graphql-config/node_modules/typescript-json-schema/dist/typescript-json-schema.js:787
                        throw new Error("Not supported: IndexSignatureDeclaration with index symbol other than a number or a string");
                              ^

Error: Not supported: IndexSignatureDeclaration with index symbol other than a number or a string

```

**Review by n1ru4l — APPROVED** [review]
Looks good, but I kind fo think we should have a test suite in the future

File: config-schema.json (hunk 1 above)
Comment (ardatan) [reply config-schema.json:138]: But we will lose the typesafety we had before in this case. 
https://github.com/kamilkisiela/graphql-config/pull/1011/files#diff-df4af8dbe3d6fbc388b8276ca4494407f03a6723884ee3d1f7e03f754aa241beL124

Comment (B2o5T) [reply config-schema.json]: @ardatan seems like `typescript-json-schema` doesn't play well with the `Record` type, I made rollback to use interfaces with index signature https://github.com/YousefED/typescript-json-schema/issues/337

Could you re-review it? 🙂

**Pull request merged**